tls/s2n_server_new_session_ticket.c

```diff
@@ -74,3 +75,112 @@ int s2n_server_nst_send(struct s2n_connection *conn)
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&stuffer, random_data->size));
+    GUARD_AS_RESULT(s2n_stuffer_read_uint32(&stuffer, ticket_age_add));
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_tls13_server_nst_send(struct s2n_connection *conn)
+{
+    notnull_check(conn);
+
+    /* Write message type because session resumption in TLS13 is a post-handshake message */
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, TLS_SERVER_NEW_SESSION_TICKET));
+
+    struct s2n_stuffer_reservation message_size = { 0 };
+    GUARD(s2n_stuffer_reserve_uint24(&conn->handshake.io, &message_size));
+
+    uint32_t ticket_lifetime_in_secs = 0;
+    GUARD_AS_POSIX(s2n_generate_ticket_lifetime(conn, &ticket_lifetime_in_secs));
+    GUARD(s2n_stuffer_write_uint32(&conn->handshake.io, ticket_lifetime_in_secs));
+
+    /* Get random data to use as ticket_age_add value */
+    struct s2n_ticket_fields ticket_fields = { 0 };
+    uint8_t data[sizeof(uint32_t)] = { 0 };
+    struct s2n_blob random_data = { 0 };
+    GUARD(s2n_blob_init(&random_data, data, sizeof(data)));
+    /** 
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.6.1
+     *#  The server MUST generate a fresh value
+     *#  for each ticket it sends.
+     **/
+    GUARD_AS_POSIX(s2n_get_private_random_data(&random_data));
+    GUARD_AS_POSIX(s2n_generate_ticket_age_add(&random_data, &ticket_fields.ticket_age_add));
+    GUARD(s2n_stuffer_write_uint32(&conn->handshake.io, ticket_fields.ticket_age_add));
+
+    /* Write ticket nonce */
+    uint8_t nonce_data[sizeof(uint16_t)] = { 0 };
+    struct s2n_blob nonce = { 0 };
+    GUARD(s2n_blob_init(&nonce, nonce_data, sizeof(nonce_data)));
+    GUARD_AS_POSIX(s2n_generate_ticket_nonce(conn->tickets_sent, &nonce));
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, sizeof(nonce_data)));
+    GUARD(s2n_stuffer_write_bytes(&conn->handshake.io, nonce.data, sizeof(nonce_data)));
+
+    /* Derive individual session ticket secret */
+    s2n_tls13_connection_keys(secrets, conn);
+    struct s2n_blob master_secret = { 0 };
+    GUARD(s2n_blob_init(&master_secret, conn->resumption_master_secret, sizeof(conn->resumption_master_secret)));
+    uint8_t session_secret_data[S2N_TLS13_SECRET_MAX_LEN] = { 0 };
+    GUARD(s2n_blob_init(&ticket_fields.session_secret, session_secret_data, secrets.size));
+    GUARD_AS_POSIX(s2n_tls13_derive_session_ticket_secret(&secrets, &master_secret, &nonce, &ticket_fields.session_secret));
+
+    /* Create ticket */
+    uint8_t ticket_data[S2N_MAX_TICKET_SIZE_IN_BYTES] = { 0 };
+    struct s2n_blob ticket_blob = { 0 };
+    struct s2n_stuffer session_ticket = { 0 };
+    GUARD(s2n_blob_init(&ticket_blob, ticket_data, sizeof(ticket_data)));
+    GUARD(s2n_stuffer_init(&session_ticket, &ticket_blob));
+    GUARD(s2n_encrypt_session_ticket(conn, &ticket_fields, &session_ticket));
+
+    /* Write session ticket */
+    GUARD(s2n_stuffer_write_uint8(&conn->handshake.io, s2n_stuffer_data_available(&session_ticket)));
```

Comment:
I know the largest a session_ticket could be is S2N_MAX_TICKET_SIZE_IN_BYTES bc that's the size of the underlying memory you gave the stuffer, but we still might want a sanity check that s2n_stuffer_data_available(&session_ticket) <= UINT8_MAX since we're technically still putting a size_t in a uint8_t.